File: arsenal/google-mcp/genai-toolbox/internal/server/static/js/toolsets.js
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { loadTools } from "./loadTools.js";

document.addEventListener('DOMContentLoaded', () => {
    const searchInput = document.getElementById('toolset-search-input');
    const searchButton = document.getElementById('toolset-search-button');
    const secondNavContent = document.getElementById('secondary-panel-content');
    const toolDisplayArea = document.getElementById('tool-display-area');

    if (!searchInput || !searchButton || !secondNavContent || !toolDisplayArea) {
        console.error('Required DOM elements not found.');
        return;
    }

    // Event listener for search button click
    searchButton.addEventListener('click', () => {
        toolDisplayArea.innerHTML = '';
        const toolsetName = searchInput.value.trim();
        if (toolsetName) {
            loadTools(secondNavContent, toolDisplayArea, toolsetName)
        } else {
            secondNavContent.innerHTML = '<p>Please enter a toolset name to see available tools. <br><br>To view the default toolset that consists of all tools, please select the "Tools" tab.</p>';
        }
    });

    // Event listener for Enter key in search input
    searchInput.addEventListener('keypress', (event) => {
        toolDisplayArea.innerHTML = '';
        if (event.key === 'Enter') {
            const toolsetName = searchInput.value.trim();
            if (toolsetName) {
                loadTools(secondNavContent, toolDisplayArea, toolsetName);
            } else {
                secondNavContent.innerHTML = '<p>Please enter a toolset name to see available tools. <br><br>To view the default toolset that consists of all tools, please select the "Tools" tab.</p>';
            }
        }
    });
})

File: arsenal/google-mcp/genai-toolbox/internal/tools/toolsets.go
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tools

import (
	"fmt"
	"regexp"
)

type ToolsetConfig struct {
	Name      string   `yaml:"name"`
	ToolNames []string `yaml:",inline"`
}

type Toolset struct {
	ToolsetConfig
	Tools       []*Tool         `yaml:",inline"`
	Manifest    ToolsetManifest `yaml:",inline"`
	McpManifest []McpManifest   `yaml:",inline"`
}

func (t Toolset) ToConfig() ToolsetConfig {
	return t.ToolsetConfig
}

type ToolsetManifest struct {
	ServerVersion string              `json:"serverVersion"`
	ToolsManifest map[string]Manifest `json:"tools"`
}

func (t ToolsetConfig) Initialize(serverVersion string, toolsMap map[string]Tool) (Toolset, error) {
	// finish toolset setup
	// Check each declared tool name exists
	var toolset Toolset
	toolset.Name = t.Name
	if !IsValidName(toolset.Name) {
		return toolset, fmt.Errorf("invalid toolset name: %s", toolset.Name)
	}
	toolset.Tools = make([]*Tool, 0, len(t.ToolNames))
	toolset.Manifest = ToolsetManifest{
		ServerVersion: serverVersion,
		ToolsManifest: make(map[string]Manifest),
	}
	for _, toolName := range t.ToolNames {
		tool, ok := toolsMap[toolName]
		if !ok {
			return toolset, fmt.Errorf("tool does not exist: %s", toolName)
		}
		toolset.Tools = append(toolset.Tools, &tool)
		toolset.Manifest.ToolsManifest[toolName] = tool.Manifest()
		toolset.McpManifest = append(toolset.McpManifest, tool.McpManifest())
	}

	return toolset, nil
}

var validName = regexp.MustCompile(`^[a-zA-Z0-9_-]*$`)

func IsValidName(s string) bool {
	return validName.MatchString(s)
}
